ui/objects/semantic-ui/js/objects/editform/components/VocabularyTreeSelectField.jsx

```diff
@@ -0,0 +1,428 @@
+        moveKey(index, null, true);
+      }
+    } else if (e.key === "ArrowRight") {
+      if (index < amount - 1) {
+        const nextColumnOptions = hierarchicalData[index + 1];
+        if (nextColumnOptions) {
+          const nextColumnIndex = nextColumnOptions.findIndex(
+            (o) => o.hierarchy.ancestors[0] === parentsState[index]
+          );
+          if (nextColumnIndex !== -1) {
+            newIndex = nextColumnIndex;
+            updateHierarchy(
+              nextColumnOptions[nextColumnIndex].value,
+              index + 1
+            )();
+            moveKey(index + 1, newIndex, false);
+          }
+        }
+      }
+    } else if (e.key === "Enter") {
+      handleSelect(data[keybState[index]], index, e);
+    } else if ((e.key = " ")) {
+      handleSelect(data[keybState[index]], index, e);
+    }
+  };
+
+  const renderColumns = (index) => {
+    return (
+      <Grid.Column>
+        {serializedOptions.map((option, i) => {
+          if (
+            option.hierarchy.ancestors.length == index &&
+            (index == 0 ||
+              option.hierarchy.ancestors[0] == parentsState[index - 1])
+          ) {
+            return (
+              <Grid.Row
+                key={i}
+                className={
+                  option.value == parentsState[index] ? "open spaced" : "spaced"
+                }
+              >
+                {multiple && (
+                  <Checkbox
+                    checked={
+                      selectedState.findIndex(
+                        (item) => item.value === option.value
+                      ) !== -1
+                    }
+                    indeterminate={
+                      selectedState.some((item) =>
+                        item.hierarchy.ancestors.includes(option.value)
+                      )
+                        ? true
+                        : false
+                    }
+                    onChange={(e) => {
+                      handleSelect(option, option.value, e);
+                    }}
+                  />
+                )}
+                <Button
+                  basic
+                  color="black"
+                  onClick={updateHierarchy(option.value, index)}
+                  onDoubleClick={(e) => {
+                    handleSelect(option, option.value, e);
+                  }}
+                  onKeyDown={(e) => {
+                    handleKey(e, option, index);
+                  }}
+                  tabIndex={0}
+                >
+                  {option.text}
+                </Button>
+                {option.element_type == "parent" && (
+                  <Button onClick={updateHierarchy(option.value, index)}>
+                    {index !== amount - 1 && <Icon name="angle right black " />}
+                  </Button>
+                )}
+              </Grid.Row>
+            );
+          }
+        })}
+      </Grid.Column>
+    );
+  };
+
+  const SearchComponent = ({ vocab }) => {
```

Comment:
Extract this component outside of Vocab select field, components shouldn't define nested components inside itself, more so if the parent component is already quite complicated